ha/monitor/k8s/parser.py

```diff
@@ -103,10 +106,23 @@ def parse(self, an_event, cached_state):
         alert.resource_type = self._type
         alert.timestamp = str(int(time.time()))
 
+        # Get value of machine id key (path of the machine id in k8s event)
+        machine_id_key = Conf.get(const.HA_GLOBAL_INDEX, f"MONITOR{_DELIM}machine_id_key")
+
+        # loop over keys and check if exist then get the value.
+        # this code is flexible can be used for any key in event for example
+        # if value at the place event[raw_object][metadat][name] then input key will be 'metadata/name'
+        # note if the key is fixed cannot chnage then can use constant here also insted of parsing
+        machine_id_keys = machine_id_key.split('/')
+        machine_id = an_event[K8SEventsConst.RAW_OBJECT]
+        for key in machine_id_keys:
+            if key != None and isinstance(machine_id, dict) and key in machine_id:
```

Comment:
Does this logic ensure that only the correct machine id will be populated in alert.resource_name? 
What happens if the string given in machine_id_key is present in any other path in raw_object, alongwith the current location ?

Reply:
it will locate the correct path only.

as python doesn't support the duplicate keys the code will parse the exact same path.

E.g.

metadata/machineId

metadata/podInfo/machineId

status/podInfo/machineId



if we provided key metadata/podInfo/machineId

then first it will search metadata then enter into metata dict.

metadata will be only 1 as python doesn't support the duplicate keys

then it will search next key podInfo again it will be only 1 and hence it will enter info the dict of podInfo.

then it will search machineId in same podInfo not other podInfo dict which is inside status.

and again here in podInfo dict machineId will be only 1 and hence reach to correft location.

Reply:
I mean what happens in case we have keys like 
metadata/machineId
status/podinfo/metadata/machineid
etc

Reply:
it will reach only the path is provided in input keys no matter how many similar keys exists

Reply:
below not is also added in code:

        # Imp Note: Below logic is required only if we are getting the absolute path where the machine-id exists in the event.
        #           i.e. metadata/podInfo/machineId so it will be parsed like [raw_object][metadata][podInfo][machineId]
        #           because k8s event we receive is multi-level nested dict and  the key can occur at multiple places
        #           for different reasons as other many keys getting repeated so choosing required key will be difficult.
        #           OR if once the key is added while creating pod and the fixed exact location in the event is found then,
        #           the changes added in /k8s_setup/ha_setup.py for this key and the below logic for parsing this key
        #           is also not required can add constant and directly fetch the value.